### PR TITLE
[NNAPI QDQ] AddQDQAdd/Mul, update to NNAPI QDQ handling, update some test settings

### PR DESCRIPTION
**Pull request by gwang-msft**
**Description**: [NNAPI QDQ] AddQDQAdd/Mul, update to NNAPI QDQ handling, update some test settings

**Motivation and Context**
- Add QDQ Add/Mul for NNAPI
- Update the NNAPI input/output checking, consolidated some helper functions
- Update the test cases to generate NNAPI specific QDQ models
- Update the NNAPI test to have optional verification for entire graph taken by EP
- Added options for `RunAndVerifyOutputsWithEP` to verify fp32 results with optional tolerance
- Fixed a small issue in previous QDQ AvergeagePool PR

### CODE REVIEW

File: onnxruntime/test/providers/nnapi/nnapi_basic_test.cc [hunk 1]
```diff
@@ -316,14 +319,38 @@ TEST(NnapiExecutionProviderTest, TestQDQResize) {
...
+                      {1, 23, 13, 13} /* input_shape */,
+                      "Add" /* op_type */),
+                  "nnapi_qdq_test_graph_add",
+                  {true /* verify_entire_graph_use_ep */});
+
+  RunQDQModelTest(BuildBinaryOpTestCase<uint8_t /* Input1Type */,
+                                        uint8_t /* Input2Type */,
+                                        uint8_t /* OutputType */>(
+                      {1, 23, 13, 13} /* input_shape */,
+                      "Add" /* op_type */),
```
Comment (YUNQIUGUO) [inline onnxruntime/test/providers/nnapi/nnapi_basic_test.cc:351]: "Add" -> "Mul"

File: onnxruntime/test/optimizer/qdq_test_utils.h [hunk 2]
```diff
@@ -110,5 +127,65 @@ GetQDQTestCaseFn BuildQDQResizeTestCase(const std::vector<int64_t>& input_shape,
...
+template <typename Input1Type, typename Input2Type, typename OutputType>
+GetQDQTestCaseFn BuildBinaryOpTestCase(const std::vector<int64_t>& input_shape,
+                                       const std::string& op_type) {
+  return [input_shape, op_type](ModelTestBuilder& builder) {
+    auto* input1_arg = builder.MakeInput<float>(input_shape, -1.f, 1.f);
+    auto* input2_arg = builder.MakeInput<float>(input_shape, -1.f, 1.f);
+    auto* output_arg = builder.MakeOutput();
+
+#ifdef USE_NNAPI  // NNAPI require consistent scales for DQ -> bin op -> Q
+    float q_scale = 0.008f;
```
Comment (skottmckay) [inline onnxruntime/test/optimizer/qdq_test_utils.h:139]: Is this the case for all binary ops? Naively I would have expected mul or div to potentially change the scale but maybe that isn't the case.

Comment (skottmckay) [reply onnxruntime/test/optimizer/qdq_test_utils.h:139]: Here's a snippet from a QDQ model exported from pytorch where the scale and zero point changes across an Add node.

image

Comment (gwang-msft) [reply onnxruntime/test/optimizer/qdq_test_utils.h:139]: Yes, the output scale and ZP can be changed, updated the test case, it is only required to be the same for QD->bin_op_input and bin_op_output->Q

File: onnxruntime/core/providers/nnapi/nnapi_builtin/builders/op_support_checker.cc [hunk 3]
```diff
@@ -58,18 +72,198 @@ bool HasExternalInitializer(const InitializedTensorSet& initializers, const Node
...
+  const auto scale_name = io_def.quant_param->scale.Name();
+  if (!Contains(initializers, scale_name)) {
+    LOGS_DEFAULT(VERBOSE) << "The scale of " << op_type << " must be an initializer tensor";
+    return false;
+  }
 
-  op_registrations.support_checkers.push_back(std::make_unique<T>());
-  for (const auto& op : op_types) {
-    op_registrations.op_support_checker_map.emplace(op, op_registrations.support_checkers.back().get());
+  const auto& scale_tensor = *initializers.at(scale_name);
```
Comment (edgchen1) [inline onnxruntime/core/providers/nnapi/nnapi_builtin/builders/op_support_checker.cc:87]: >at [](http://example.com/codeflow?start=43&length=2)

general comment: double lookup of `Contains(container, key)` then `container.at(key)` can be changed to `auto it = container.find(key); it != container.end()` and `it->second`

File: onnxruntime/core/providers/nnapi/nnapi_builtin/builders/op_support_checker.cc [hunk 4]
```diff
@@ -58,18 +72,198 @@ bool HasExternalInitializer(const InitializedTensorSet& initializers, const Node
...
+
+    // We only support u8 for most of the inputs and all outputs, with the exception for Quantized MatMul and Conv,
+    // which allows s8 weight (u8s8)
+    // TODO, add support of s8s8
+    if (input_type != ONNX_NAMESPACE::TensorProto_DataType_UINT8 &&
+        !(input_type == ONNX_NAMESPACE::TensorProto_DataType_INT8 && is_conv_matmul_u8s8_weight)) {
+      LOGS_DEFAULT(VERBOSE) << op_type << "NodeUnit [" << node_unit.Name()
+                            << "], type [" << op_type << "]'s "
+                            << (is_input ? "Input" : "Output") << " index  [" << idx
+                            << "] has unspoorted type [" << input_type << "]";
```
Comment (edgchen1) [inline onnxruntime/core/providers/nnapi/nnapi_builtin/builders/op_support_checker.cc:250]: >unspoorted [](http://example.com/codeflow?start=38&length=10)

"unsupported"

File: onnxruntime/core/providers/nnapi/nnapi_builtin/builders/op_support_checker.cc [hunk 5]
```diff
@@ -281,21 +505,29 @@ int BinaryOpSupportChecker::GetMinSupportedOpSet(const NodeUnit& node_unit) cons
...
+
+  if (is_quantized_op) {
+    // QLinearAdd/QDQAdd/QLinearMul/QDQMul
+    if (!HasValidBinaryOpQuantizedInputTypes(node_unit))
+      return false;
 
-  if (is_qlinear_add) {
-    // QLinearAdd
-    if (!HasValidBinaryOpQuantizedInputs(node_unit))
+    if (!IsQuantizedIOSupported(initializers, node_unit, {0, 1}, params, true /* is_input */))
```
Comment (edgchen1) [inline onnxruntime/core/providers/nnapi/nnapi_builtin/builders/op_support_checker.cc:527]: >/* is_input */ [](http://example.com/codeflow?start=78&length=14)

just a thought - if it was an enum, say `enum class IOKind { Input, Output }`, we wouldn't need comments

File: onnxruntime/core/providers/nnapi/nnapi_builtin/builders/helper.cc [hunk 6]
```diff
@@ -146,16 +140,17 @@ bool HasValidBinaryOpQuantizedInputs(const NodeUnit& node_unit) {
   if (!GetType(inputs[1].node_arg, b_input_type))
     return false;
 
-  // QlinearConv supports u8u8 or u8s8
-  // QLinearMatMul/Add only support u8u8
-  bool is_quant_conv = IsQuantizedConv(quant_op_type);
+  // QlinearConv/Mul supports u8u8 or u8s8
```
[the remark this answers is not in the document]
Comment (gwang-msft) [reply onnxruntime/core/providers/nnapi/nnapi_builtin/builders/helper.cc:143]: Yes, it should be MatMul

File: onnxruntime/core/providers/nnapi/nnapi_builtin/builders/op_support_checker.cc (hunk 3 above)
Comment (edgchen1) [reply onnxruntime/core/providers/nnapi/nnapi_builtin/builders/op_support_checker.cc]: also shows up in HasExternalInitializer(), IsQuantizationZeroPointSupported(), ReshapeOpSupportChecker::IsOpSupportedImpl(), and ConvOpSupportChecker::IsOpSupportedImpl()
can also address it later

**Review by edgchen1 — APPROVED** [review]
:shipit:

**Pull request merged**